### PR TITLE
Feat/ipv6

### LINKED ISSUE #21: IPv6 Support
Hi,

I don't know if this is of interest, but I've added rudimentary IPv6 support in a fork of the repo:

https://github.com/everycity/terraform-aws-dynamic-subnets
https://github.com/everycity/terraform-aws-vpc

Wasn't sure what the best way to communicate this to the world was, hence raising an issue! Anyway thanks for the awesome work - really helpful template.

Cheers,

Alasdair

### PR DESCRIPTION
**Pull request by marekmoscichowski**
## what
* Enable subnets to have ipv6 addresses assigned
* Added ACL entries to allow traffic via ipv6
* Added routes. From private subnets to egress only internet gateway, From public subnets to the internet gateway

## why
* ipv6 is here. The big advantage is to have the ability to access the internet from a private subnet without the need of a costly nat-gateway. (If the target supports ipv6)

## references
closes #21

### CODE REVIEW

File: public.tf [hunk 1]
```diff
@@ -62,6 +65,18 @@ resource "aws_route" "public" {
   }
 }
 
+resource "aws_route" "public-ipv6" {
```
Comment (nitrocode) [inline public.tf:68]: dashes in resource names should be underscores
```suggestion
resource "aws_route" "public_ipv6" {
```

File: variables.tf [hunk 2]
```diff
@@ -141,3 +146,9 @@ variable "root_block_device_encrypted" {
   default     = true
   description = "Whether to encrypt the root block device"
 }
+
+variable "map_private_ipv6_on_launch" {
```
Comment (nitrocode) [inline variables.tf:150]:
```suggestion
variable "assign_ipv6_address_on_creation" {
```

File: variables.tf [hunk 3]
```diff
@@ -25,6 +25,11 @@ variable "igw_id" {
   description = "Internet Gateway ID the public route table will point to (e.g. `igw-9c26a123`)"
 }
 
+variable "egress_only_igw_id" {
+  type        = string
+  description = "Egress Only Internet Gateway ID the privete route tables will point to (e.g. `eigw-9c26a123`)"
```
Comment (nitrocode) [inline variables.tf:30]: This var should have a default or it will be a required input

File: private.tf [hunk 4]
```diff
@@ -63,6 +78,15 @@ resource "aws_network_acl" "private" {
   vpc_id     = var.vpc_id
   subnet_ids = aws_subnet.private.*.id
 
+  ingress {
```
Comment (nitrocode) [inline private.tf:81]: if assign_ipv6_address_on_creation is false then these ipv6 ingress and egress shouldn't be set. these can be disabled using a dynamic

File: public.tf [hunk 5]
```diff
@@ -88,6 +103,15 @@ resource "aws_network_acl" "public" {
     protocol   = "-1"
   }
 
+  egress {
```
Comment (nitrocode) [inline public.tf:106]: use dynamic here too

File: public.tf [hunk 6]
```diff
@@ -97,5 +121,14 @@ resource "aws_network_acl" "public" {
     protocol   = "-1"
   }
 
+  ingress {
```
Comment (nitrocode) [inline public.tf:124]: dynamic

File: private.tf [hunk 7]
```diff
@@ -72,6 +96,15 @@ resource "aws_network_acl" "private" {
     protocol   = "-1"
   }
 
+  egress {
```
Comment (nitrocode) [inline private.tf:99]: dynamic

File: private.tf [hunk 8]
```diff
@@ -52,6 +55,18 @@ resource "aws_route_table" "private" {
   )
 }
 
+resource "aws_route" "private-ipv6" {
```
Comment (nitrocode) [inline private.tf:58]:
```suggestion
resource "aws_route" "private_ipv6" {
```

File: private.tf [hunk 9]
```diff
@@ -52,6 +55,18 @@ resource "aws_route_table" "private" {
...
+  count                       = local.enabled ? local.availability_zones_count : 0
+  route_table_id              = aws_route_table.private[count.index].id
+  destination_ipv6_cidr_block = "::/0"
```
Comment (nitrocode) [inline private.tf:61]: if assign_ipv6_address_on_creation is false, shouldn't this be disabled

File: public.tf [hunk 10]
```diff
@@ -62,6 +65,18 @@ resource "aws_route" "public" {
   }
 }
 
+resource "aws_route" "public-ipv6" {
+  count                       = local.public_route_expr_enabled ? 0 : local.enabled_count
```
Comment (nitrocode) [inline public.tf:69]: this should be disabled if assign_ipv6_address_on_creation = false

**Comment by mihaiplesa** [comment]
@marekmoscichowski is this still maintained by you?

**Comment by carsondoesbusiness** [comment]
Any plans to merge this soon? Really need this feature!

**Comment by aknysh** [comment]
@marekmoscichowski  please address the comments by @nitrocode

**Comment by marekmoscichowski** [comment]
Hi, @aknysh @mihaiplesa 

unfortunately, I've got a lot on my plate atm. But I saw that there is a similar pr opened quite recently which might be currently maintained  https://github.com/cloudposse/terraform-aws-dynamic-subnets/pull/147

**Comment by amexboy** [comment]
@marekmoscichowski  if you could add me as a writer on your repository I could maintain this PR! I think it would be faster than requesting a new review!

**Pull request closed without merging**